ble/src/main/java/no/nordicsemi/android/ble/MainThreadBluetoothGattCallback.kt

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this
+ * software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package no.nordicsemi.android.ble
+
+import android.bluetooth.BluetoothGatt
+import android.bluetooth.BluetoothGattCallback
+import android.bluetooth.BluetoothGattCharacteristic
+import android.bluetooth.BluetoothGattDescriptor
+import android.os.Build
+import android.os.Handler
+
+import androidx.annotation.IntRange
+import androidx.annotation.Keep
+import androidx.annotation.RequiresApi
+
+import no.nordicsemi.android.ble.annotation.ConnectionState
+import no.nordicsemi.android.ble.annotation.PhyValue
+
+/**
+ * This class ensures that the BLE callbacks will be called on the main (UI) thread.
+ * Handler parameter was added to [ #connectGatt(Context, boolean, BluetoothGattCallback, int, int, Handler)][android.bluetooth.BluetoothDevice]
+ * in Android Oreo, before that the behavior was undefined.
+ */
+abstract class MainThreadBluetoothGattCallback : BluetoothGattCallback() {
+    private var mHandler: Handler? = Handler()
```

Comment:
```suggestion
    private var mHandler: Handler? = Handler(Looper.getMainLooper())
```

Argumentless Handler() constructor uses Looper from currently running thread. This means that if this object is created from a background thread, the runnables posted to the handler are executed on that thread, not main thread.

Reply:
Very good observation, thank you.